File: bench/python/bench_survts.py
#!/usr/bin/env python3
"""Survival / time-series / quantile / GAM / spline benchmarks (B8) — Python side.

Compares against statsmodels (ARIMA, quantile regression), lifelines
(Cox PH, Kaplan-Meier), pygam (GAM), and scipy.interpolate (PCHIP).

Run with::

    OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 \\
        bench/venv/bin/python bench/python/bench_survts.py

Writes ``bench/results/python/survts.csv``.
"""

from __future__ import annotations

import csv
import random
import time
from dataclasses import dataclass
from pathlib import Path

import numpy as np
import pandas as pd


REPO = Path(__file__).resolve().parents[2]
DATA = REPO / "bench" / "data"
OUT = REPO / "bench" / "results" / "python"
OUT.mkdir(parents=True, exist_ok=True)


@dataclass
class Row:
    name: str
    time_ms: float
    acc_main: float
    acc_aux: float
    extra: str


def median_time(fn, n_iter: int) -> tuple[float, object]:
    times = []
    last = None
    for _ in range(n_iter):
        t0 = time.perf_counter()
        last = fn()
        t1 = time.perf_counter()
        times.append(1000.0 * (t1 - t0))
    times.sort()
    return times[len(times) // 2], last


# ---------------------------------------------------------------------------
# Synthetic data (deterministic seeds matching Haskell bench)
# ---------------------------------------------------------------------------

def gen_ar1(n: int) -> np.ndarray:
    rng = random.Random(42)
    xs = []
    x = 0.0
    while len(xs) < n:
        z = rng.uniform(-3.0, 3.0)
        x = 0.7 * x + 0.3 * z
        xs.append(x)
    return np.array(xs)


def gen_surv(n: int) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    rng = random.Random(7)
    rows_x, rows_t, rows_e = [], [], []
    for _ in range(n):
        x1 = rng.uniform(-1.0, 1.0)
        x2 = rng.uniform(-1.0, 1.0)
        u = rng.uniform(0.01, 1.0)
        t = -np.log(u) / np.exp(0.5 * x1 - 0.3 * x2)
        c = rng.uniform(0.0, 1.0)
        ev = 1 if c < 0.7 else 0
        rows_x.append([x1, x2])
        rows_t.append(t)
        rows_e.append(ev)
    return (np.array(rows_x), np.array(rows_t), np.array(rows_e))


# ---------------------------------------------------------------------------
# ARIMA(1,1,1)
# ---------------------------------------------------------------------------

def bench_arima(name: str, n_iter: int = 3) -> Row:
    from statsmodels.tsa.arima.model import ARIMA

    y = gen_ar1(1000)

    def run():
        return ARIMA(y, order=(1, 1, 1)).fit(method_kwargs={"disp": 0})

    ms, fit = median_time(run, n_iter)
    return Row(name, ms, 0, 0,
               "statsmodels.tsa.arima.ARIMA p=1 d=1 q=1 n=1000")


# ---------------------------------------------------------------------------
# Cox PH
# ---------------------------------------------------------------------------

def bench_coxph(name: str, n_iter: int = 5) -> Row:
    from lifelines import CoxPHFitter

    x, t, e = gen_surv(2000)
    df = pd.DataFrame({"x1": x[:, 0], "x2": x[:, 1], "T": t, "E": e})

    def run():
        cph = CoxPHFitter()
        cph.fit(df, duration_col="T", event_col="E")
        return cph

    ms, fit = median_time(run, n_iter)
    b = fit.params_
    return Row(name, ms, float(b["x1"]), float(b["x2"]),
               "lifelines.CoxPHFitter n=2000 p=2 30% censor")


# ---------------------------------------------------------------------------
# Kaplan-Meier
# ---------------------------------------------------------------------------

def bench_km(name: str, n_iter: int = 5) -> Row:
    from lifelines import KaplanMeierFitter

    _, t, e = gen_surv(2000)

    def run():
        km = KaplanMeierFitter()
        km.fit(t, event_observed=e)
        return km

    ms, km = median_time(run, n_iter)
    sf = km.survival_function_["KM_estimate"]
    t_end = float(sf.index[-1])
    s_end = float(sf.iloc[-1])
    return Row(name, ms, t_end, s_end,
               "lifelines.KaplanMeierFitter n=2000")


# ---------------------------------------------------------------------------
# Quantile regression
# ---------------------------------------------------------------------------

def bench_quantile(name: str, n_iter: int = 3) -> Row:
    from statsmodels.regression.quantile_regression import QuantReg

    df = pd.read_csv(DATA / "lm_n10000_p50.csv")
    x = df.iloc[:, :20].to_numpy()
    y = df.iloc[:, -1].to_numpy()
    # Add intercept column.
    X = np.column_stack([np.ones(len(y)), x])

    def run():
        return QuantReg(y, X).fit(q=0.5)

    ms, fit = median_time(run, n_iter)
    return Row(name, ms, 0, 0,
               "statsmodels QuantReg tau=0.5 n=10000 p=20")


# ---------------------------------------------------------------------------
# GAM
# ---------------------------------------------------------------------------

def bench_gam(name: str, n_iter: int = 3) -> Row:
    from pygam import LinearGAM, s

    df = pd.read_csv(DATA / "kernel_n2000_p5.csv")
    x = df.iloc[:, :2].to_numpy()
    y = df.iloc[:, -1].to_numpy()

    def run():
        gam = LinearGAM(s(0, n_splines=8) + s(1, n_splines=8))
        return gam.fit(x, y)

    ms, fit = median_time(run, n_iter)
    return Row(name, ms, 0, 0,
               "pygam.LinearGAM n_splines=8 n=2000 p=2")


# ---------------------------------------------------------------------------
# 1D spline (PCHIP)
# ---------------------------------------------------------------------------

def bench_spline(name: str, n_iter: int = 5) -> Row:
    from scipy.interpolate import PchipInterpolator

    n = 1000
    xs = np.linspace(0, 1, n)
    ys = np.sin(3 * xs) + 0.1 * np.cos(15 * xs)
    qs = np.linspace(0, 1, 5000)

    def run():
        f = PchipInterpolator(xs, ys)
        return float(f(qs).sum())

    ms, total = median_time(run, n_iter)
    return Row(name, ms, total, 0,
               "scipy.interpolate.PchipInterpolator build n=1000 + eval @5000")


# ---------------------------------------------------------------------------

def write_rows(path: Path, rows: list[Row], suite: str = "survts") -> None:
    with path.open("w", newline="") as fh:
        w = csv.writer(fh)
        w.writerow(["system", "suite", "name",
                    "time_ms", "acc_main", "acc_aux", "extra"])
        for r in rows:
            w.writerow(["python", suite, r.name,
                        f"{r.time_ms:.6g}", f"{r.acc_main:.6g}",
                        f"{r.acc_aux:.6g}", r.extra])


def main() -> None:
    rows: list[Row] = []
    for fn, name in [
        (bench_arima,    "ARIMA_n1000_pdq111"),
        (bench_coxph,    "CoxPH_n2000_p2_30pct_censor"),
        (bench_km,       "KM_n2000"),
        (bench_quantile, "Quantile_n10000_p20_tau0.5"),
        (bench_gam,      "GAM_n2000_p2_d3_k5"),
        (bench_spline,   "Spline_PCHIP_n1000"),
    ]:
        print(f"Running {name} …")
        rows.append(fn(name))
    out = OUT / "survts.csv"
    write_rows(out, rows)
    print(f"wrote {len(rows)} rows → {out}")


if __name__ == "__main__":
    main()

File: bench/haskell/BenchSurvTS.hs
{-# LANGUAGE OverloadedStrings #-}
{-# OPTIONS_GHC -fno-full-laziness -fno-cse #-}
-- | Survival / time-series / quantile / GAM / spline benchmarks (B8).
--
-- Compares hanalyze against statsmodels (ARIMA, quantile regression),
-- lifelines (Cox PH, Kaplan-Meier), pygam (GAM), and scipy.interpolate
-- (1D spline interpolation).
--
-- Outputs the unified BenchRow CSV at @bench/results/haskell/survts.csv@.
module Main where

import qualified Data.Vector             as V
import qualified Numeric.LinearAlgebra   as LA
import           System.Random           (mkStdGen, randomR)

import qualified Model.TimeSeries        as TS
import qualified Model.Survival          as Surv
import qualified Model.Quantile          as QR
import qualified Model.GAM               as GAM
import qualified Stat.Interpolate        as Interp

import           BenchUtil

-- ---------------------------------------------------------------------------
-- Synthetic data generators (deterministic seeds)
-- ---------------------------------------------------------------------------

-- | AR(1) series with phi=0.7 and Gaussian noise.
genAR1 :: Int -> LA.Vector Double
genAR1 n = LA.fromList (go (mkStdGen 42) 0.0 [])
  where
    go _ _ acc | length acc >= n = reverse acc
    go g x acc =
      let (z, g') = randomR (-3.0, 3.0) g
          x'     = 0.7 * x + 0.3 * z
      in go g' x' (x' : acc)

-- | Survival data: exponential time + 30% censoring.
genSurv :: Int -> ([LA.Vector Double], [Surv.SurvSample])
genSurv n =
  let g0 = mkStdGen 7
      (rows, _) = foldr step ([], g0) [1 .. n]
      step _ (acc, g) =
        let (x1, g1) = randomR (-1.0 :: Double, 1.0) g
            (x2, g2) = randomR (-1.0 :: Double, 1.0) g1
            (u,  g3) = randomR (0.01 :: Double, 1.0) g2
            t        = -log u / exp (0.5 * x1 - 0.3 * x2)
            (c,  g4) = randomR (0.0 :: Double, 1.0) g3
            ev       = if c < 0.7 then Surv.Observed else Surv.Censored
        in ((LA.fromList [x1, x2], Surv.SurvSample t ev) : acc, g4)
  in unzip rows

-- ---------------------------------------------------------------------------

main :: IO ()
main = do
  rows <- mconcat <$> sequence
    [ benchARIMA   "ARIMA_n1000_pdq111"
    , benchCoxPH   "CoxPH_n2000_p2_30pct_censor"
    , benchKM      "KM_n2000"
    , benchQuant   "Quantile_n10000_p20_tau0.5"
    , benchGAM     "GAM_n2000_p2_d3_k5"
    , benchSpline  "Spline_PCHIP_n1000"
    ]
  writeRows "bench/results/haskell/survts.csv" rows
  putStrLn $ "wrote " ++ show (length rows)
          ++ " rows → bench/results/haskell/survts.csv"

-- ---------------------------------------------------------------------------
-- ARIMA(1,1,1)
-- ---------------------------------------------------------------------------

benchARIMA :: String -> IO [BenchRow]
benchARIMA name = do
  let y = genAR1 1000
      run :: Int -> IO TS.ARIMAFit
      run _ = pure $! TS.fitARIMA 1 1 1 y
  (ms, fit) <- timeitTastyIO probe run
  return [ BenchRow "haskell" "survts" name ms 0 0
            ("Model.TimeSeries.fitARIMA p=1 d=1 q=1 n=1000") ]
  where
    probe f = LA.sumElements (TS.forecastARIMA f 10)

-- ---------------------------------------------------------------------------
-- Cox PH
-- ---------------------------------------------------------------------------

benchCoxPH :: String -> IO [BenchRow]
benchCoxPH name = do
  let (xs, samples) = genSurv 2000
      run :: Int -> IO Surv.CoxFit
      run _ = pure $! Surv.coxPH xs samples
  (ms, fit) <- timeitTastyIO probe run
  let beta = Surv.coxBeta fit
      b1 = beta LA.! 0
      b2 = beta LA.! 1
  return [ BenchRow "haskell" "survts" name ms b1 b2
            ("Model.Survival.coxPH n=2000 p=2 (Newton-Raphson)") ]
  where
    probe f = LA.sumElements (Surv.coxBeta f)

-- ---------------------------------------------------------------------------
-- Kaplan-Meier
-- ---------------------------------------------------------------------------

benchKM :: String -> IO [BenchRow]
benchKM name = do
  let (_, samples) = genSurv 2000
      run :: Int -> IO Surv.KMResult
      run _ = pure $! Surv.kaplanMeier samples
  (ms, res) <- timeitTastyIO probe run
  let ts   = Surv.kmrTimes res
      surv = Surv.kmrSurvival res
      tEnd = if null ts then 0 else last ts
      sEnd = if null surv then 1 else last surv
  return [ BenchRow "haskell" "survts" name ms tEnd sEnd
            ("Model.Survival.kaplanMeier n=2000") ]
  where
    probe r = sum (Surv.kmrSurvival r)

-- ---------------------------------------------------------------------------
-- Quantile regression (median, tau=0.5)
-- ---------------------------------------------------------------------------

benchQuant :: String -> IO [BenchRow]
benchQuant name = do
  (x, y) <- readCsvXY "bench/data/lm_n10000_p50.csv"
  -- Use first 20 columns for fair comparison with Python.
  let xCut = LA.takeColumns 20 x
      run :: Int -> IO QR.QRFit
      run _ = pure $! QR.fitQuantile 0.5 xCut y
  (ms, fit) <- timeitTastyIO probe run
  return [ BenchRow "haskell" "survts" name ms 0 0
            ("Model.Quantile.fitQuantile tau=0.5 n=10000 p=20") ]
  where
    probe f = LA.sumElements (QR.qfBeta f)

-- ---------------------------------------------------------------------------
-- GAM (degree=3, knots=5, two predictors)
-- ---------------------------------------------------------------------------

benchGAM :: String -> IO [BenchRow]
benchGAM name = do
  (x, y) <- readCsvXY "bench/data/kernel_n2000_p5.csv"
  let cols = LA.toColumns x
      x1   = V.fromList (LA.toList (head cols))
      x2   = V.fromList (LA.toList (cols !! 1))
      yV   = V.fromList (LA.toList y)
      run :: Int -> IO GAM.GAMFit
      run _ = pure $! GAM.fitGAM 3 5 1.0 [x1, x2] yV
  (ms, fit) <- timeitTastyIO probe run
  return [ BenchRow "haskell" "survts" name ms 0 0
            ("Model.GAM.fitGAM degree=3 knots=5 lambda=1.0 n=2000 p=2") ]
  where
    probe f = LA.sumElements (GAM.gamYHat f)

-- ---------------------------------------------------------------------------
-- 1D spline interpolation (PCHIP)
-- ---------------------------------------------------------------------------

benchSpline :: String -> IO [BenchRow]
benchSpline name = do
  let n = 1000
      xs = [fromIntegral i / fromIntegral (n - 1) | i <- [0 .. n - 1]]
      ys = map (\xi -> sin (3 * xi) + 0.1 * cos (15 * xi)) xs
      pts = zip xs ys
      f = Interp.interp1d Interp.PCHIP pts
      -- Evaluate at 5000 query points.
      qs = [fromIntegral i / 4999.0 | i <- [0 .. 4999 :: Int]]
      run :: Int -> IO Double
      run _ = pure $! sum [f q | q <- qs]
  (ms, total) <- timeitTastyIO id run
  return [ BenchRow "haskell" "survts" name ms total 0
            ("Stat.Interpolate.interp1d PCHIP, build n=1000 + eval @5000 pts") ]
